### PR TITLE
SyncComputing - Restaurant Endpoint Challenge (K.DeJan)

### PR DESCRIPTION
**Pull request by MisterDeejay**
This PR contains the major changes required to get the restaurants' endpoint functioning. I've left notes on my PR to hopefully clarify the code and choices I made, as well as, my thought process throughout the assessment. The following is a short list of tasks I would do with more time:
* Add auth by JWT
* Add more test coverage to handle more edge cases
* Change IDs in the `restaurants` table to UUIDs
* Look for more edge cases to cover by rescuing and returning human-readable errors

### CODE REVIEW

File: app/blueprints/restaurant_blueprint.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class RestaurantBlueprint < Blueprinter::Base
```
Comment (MisterDeejay) [inline app/blueprints/restaurant_blueprint.rb:1]: This class is used by `RestaurantsController` to return only restaurant names (listed as a product requirement) via the endpoint

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,7 @@
 class ApplicationController < ActionController::API
+  rescue_from(ActionController::BadRequest, with: :render_bad_request_response)
+
+  def render_bad_request_response(error)
+    render(json: { errors: [{ detail: "Bad request: #{error.message}" }] }, status: :bad_request)
```
Comment (MisterDeejay) [inline app/controllers/application_controller.rb:5]: The JSON object is returned when an invalid `datetime` query param is sent through the request

File: app/controllers/restaurants_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  private
+
+  def index_params
+    params.permit(:datetime)
+  end
+
+  def set_restaurants
+    @restaurants = Restaurant.all
```
Comment (MisterDeejay) [inline app/controllers/restaurants_controller.rb:15]: All restaurants are returned unless a valid `datetime` params is sent

File: app/controllers/restaurants_controller.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    if index_params[:datetime].present?
+      datetime = DateTime.parse(index_params[:datetime])
+      @restaurants = Restaurants::Filter.call(restaurants: @restaurants, datetime: datetime)
```
Comment (MisterDeejay) [inline app/controllers/restaurants_controller.rb:18]: If a `datetime` param is present, the endpoint uses the `Restaurants::Filter` service to select only restaurants open on the given `DateTime`

File: app/controllers/restaurants_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    end
+  rescue Date::Error => _e
+    raise ActionController::BadRequest, 'datetime must be a valid date'
```
Comment (MisterDeejay) [inline app/controllers/restaurants_controller.rb:21]: If an invalid `datetime` param is sent, the error raised is rescued and the appropriate error is raised to take advantage of the code in `ApplicationController`

File: db/migrate/20220921220322_create_restaurants.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    enable_extension "fuzzystrmatch" unless extension_enabled?('fuzzystrmatch')
+    enable_extension "btree_gin" unless extension_enabled?('btree_gin')
+
+    create_table :restaurants do |t|
+      t.string :name, null: false, index: { unique: true }
+      t.hstore :hours
+      t.timestamps
+    end
+
+    add_index :restaurants, :hours, using: :gin
```
Comment (MisterDeejay) [inline db/migrate/20220921220322_create_restaurants.rb:15]: This index was originally added in the hopes the restaurants could be filtered at the database level instead of through the code in `Restaurants::Filter`; however, with the addition of that service class, this index is not necessary presently

File: db/seeds.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
+# This file should contain all the record creation needed to seed the database with its default values.
```
Comment (MisterDeejay) [inline db/seeds.rb:1]: This seeds file ingests `restaurants_times.csv` and adds each `Restaurant` to the database.
